src/3rd_party/topk.cuh

```diff
@@ -0,0 +1,302 @@
+/*
+* Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+* This code is modified from the topk implementation in NVIDIA's faster
+* transformer repository. The original source code files can be found here:
+*
+* https://github.com/NVIDIA/DeepLearningExamples/blob/master/FasterTransformer/v3.0/fastertransformer/cuda/topk_kernels.cu
+* https://github.com/NVIDIA/DeepLearningExamples/blob/master/FasterTransformer/v3.0/fastertransformer/cuda/topk_kernels.cuh
+*/
+
+#pragma once
+#include <cuda_runtime.h>
+#include <cuda_fp16.h>
+#include <cmath>
+#if CUDA_VERSION >= 11000
+#include <cub/cub.cuh>
+#else
+#include "cub/cub/cub.cuh"
+#endif
+
+#define MAX_BLOCKS_PER_BEAM 8
+
+template<typename T> 
+struct FpInfinity;
+
+template <>
+struct FpInfinity<float> {
+  static __host__ __device__ __forceinline__ float infinity() {
+      return INFINITY;
+  }
+};
+
+template <>
+struct FpInfinity<__half> {
+  static __host__ __device__ __forceinline__ __half infinity() {
+      return __float2half(INFINITY);
+  }
+};
+
+template<typename IndexType, typename T>
+struct TopK {
+  IndexType p = 0;
+  T u = -FpInfinity<T>::infinity();
+
+  __device__ __forceinline__ void insertKeepMax(T elem, IndexType elem_id) {
+    if(elem > u) {
+      u = elem;
+      p = elem_id;
+    }
+  }
+
+  __device__ __forceinline__ void insertKeepMin(T elem, IndexType elem_id) {
+    if(elem < u) {
+      u = elem;
+      p = elem_id;
+    }
+  }
+
+  __device__ __forceinline__ void init(bool descendingOrder) {
+    u = descendingOrder? -FpInfinity<T>::infinity() : FpInfinity<T>::infinity();
```

Comment:
I don't think I understand what you mean entirely. We need to initialize negative inf for descending order and positive inf for ascending. It seems that this check needs to happen at runtime.

One option that would remove the check at runtime would be to make the descending order flag a template parameter. In that case, the topk operator in topk.cu would call the correct kernel which wouldn't have this check. However, I would expect that the performance benefits to doing this are small since this is a small % of the overall execution and all threads in a warp will end up taking the same path.

If I misunderstood what you meant here please clarify!

Reply:
My question was whether we ever use it both ways, or only one way, throughout Marian. If we only ever use `descending = true`, for example, we could remove this option altogether. Probably OK to keep as is. Will close this.